File: app/src/main/java/com/kee0kai/thekey/domain/StorageFilesRepository.java
package com.kee0kai.thekey.domain;

import static com.kee0kai.thekey.App.DI;

import com.kee0kai.thekey.domain.room.KeyDatabase;
import com.kee0kai.thekey.domain.room.model.StorageFileEntry;
import com.kee0kai.thekey.model.Storage;

import java.io.File;

public class StorageFilesRepository {

    private final KeyDatabase db = DI.provider().keyDatabase();


    public Storage[] getStorages() {
        StorageFileEntry[] fileEntries = db.cachedFilesDao().get();
        Storage[] storages = new Storage[fileEntries.length];
        for (int i = 0; i < fileEntries.length; i++) {
            storages[i] = new Storage(fileEntries[i].path, fileEntries[i].name, fileEntries[i].description);
        }
        return storages;
    }

    public void addStorage(Storage storage){
        StorageFileEntry cachedStorage = db.cachedFilesDao().get(storage.path);

        StorageFileEntry fileEntry = new StorageFileEntry(storage.path, storage.name, storage.description);
        if (cachedStorage != null) fileEntry.id = cachedStorage.id;
        db.cachedFilesDao().insert(fileEntry);
    }

    public Storage findStorage(String path) {
        StorageFileEntry storageFileEntry = db.cachedFilesDao().get(path);
        return storageFileEntry != null ? new Storage(storageFileEntry.path, storageFileEntry.name, storageFileEntry.description) : null;
    }

    public void deleteStorage(String path) {
        new File(path).delete();
        db.cachedFilesDao().delete(path);
    }

}

File: app_mobile/src/main/kotlin/com/github/klee0kai/thekey/app/data/StorageFilesRepository.kt
package com.github.klee0kai.thekey.app.data

import com.github.klee0kai.stone.type.wrappers.getValue
import com.github.klee0kai.thekey.app.data.room.entry.StorageFileEntry
import com.github.klee0kai.thekey.app.di.DI
import com.github.klee0kai.thekey.app.model.Storage
import kotlinx.coroutines.async
import kotlinx.coroutines.launch
import java.io.File

class StorageFilesRepository {

    private val storagesDao by DI.storageDaoLazy()

    private val scope = DI.ioThreadScope()

    fun getStorages() = scope.async {
        storagesDao.getAll().mapNotNull { entry ->
            if (entry.path != null) {
                Storage(
                    path = entry.path,
                    name = entry.name,
                    description = entry.description
                )
            } else {
                null
            }
        }
    }

    fun findStorage(path: String) = scope.async {
        val cachedStorage = storagesDao.getAll(path)
        if (cachedStorage?.path != null) {
            Storage(
                path = cachedStorage.path,
                name = cachedStorage.name,
                description = cachedStorage.description
            )
        } else {
            null
        }
    }

    fun deleteStorage(path: String) = scope.launch {
        File(path).deleteOnExit()
        storagesDao.delete(path)
    }

    fun addStorage(storage: Storage) = scope.launch {
        val cachedStorage = storagesDao.getAll(storage.path)
        var fileEntry = StorageFileEntry(
            path = storage.path,
            name = storage.name,
            description = storage.description
        )
        if (cachedStorage != null) fileEntry = fileEntry.copy(id = cachedStorage.id)
        storagesDao.insert(fileEntry)
    }


}